File: android/app/src/main/java/com/letstrack/agent/network/NetworkClient.kt
package com.letstrack.agent.network

import com.google.gson.Gson
import io.socket.client.IO
import io.socket.client.Socket
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import org.json.JSONObject
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import retrofit2.http.*
import java.net.URISyntaxException

// ============================================
// DTO DEFINITIONS
// ============================================
data class LoginRequest(val email: String, val password: String)
data class GoogleLoginRequest(val idToken: String)

data class UserProfile(
    val id: String,
    val name: String,
    val email: String,
    val role: String,
    val status: String
)

data class TenantDetails(
    val id: String,
    val name: String,
    val domain: String,
    val apiKey: String
)

data class LoginResponse(
    val token: String,
    val user: UserProfile,
    val tenant: TenantDetails
)

data class AnalyticsResponse(
    val totalVisitors: Int,
    val onlineVisitors: Int,
    val activeConversations: Int,
    val unassignedConversations: Int,
    val totalChats: Int,
    val totalAgents: Int,
    val onlineAgents: Int
)

data class VisitorDto(
    val _id: String,
    val name: String,
    val email: String?,
    val phoneNumber: String? = null,
    val country: String,
    val city: String,
    val deviceType: String,
    val currentUrl: String?,
    val isOnline: Boolean,
    val isMuted: Boolean? = false,
    val source: String? = null,
    val channel: String? = null,
    val firstSeen: String? = null,
    val lastSeen: String? = null
) {
    val resolvedChannel: String
        get() {
            if (!channel.isNullOrEmpty()) return channel
            if (!source.isNullOrEmpty()) return source
            val low = name.lowercase()
            if (low.contains("whatsapp") || (!phoneNumber.isNullOrEmpty())) return "whatsapp"
            if (low.contains("instagram") || low.contains("ig")) return "instagram"
            if (low.contains("facebook") || low.contains("fb")) return "facebook"
            return "livechat"
        }
}

data class ConversationDto(
    val _id: String,
    val visitorId: String,
    val status: String,
    val assignedAgentId: String?,
    val channel: String? = null,
    val lastMessage: String? = null,
    val unreadCount: Int? = null,
    val updatedAt: String
) {
    val resolvedChannel: String
        get() = channel?.takeIf { it.isNotEmpty() } ?: "livechat"
}

data class MessageDto(
    val _id: String,
    val conversationId: String,
    val senderType: String,
    val senderId: String,
    val senderName: String,
    val text: String,
    val timestamp: String
)

data class FcmTokenRequest(val fcmToken: String)

data class ResetPasswordRequest(val email: String, val newPassword: String)
data class ResetPasswordResponse(val message: String)

data class QuickReplyDto(
    val _id: String,
    val tenantId: String,
    val shortcut: String,
    val text: String
)

data class UpdateProfileRequest(
    val name: String?,
    val avatarUrl: String?,
    val password: String?
)

data class RegisterAgentRequest(
    val name: String,
    val email: String,
    val password: String
)

data class RegisterAgentResponse(
    val message: String,
    val agent: UserProfile
)

// ============================================
// LEAD MANAGEMENT SYSTEM DTOs
// ============================================

data class LeadNoteDto(
    val _id: String? = null,
    val text: String,
    val authorName: String? = null,
    val createdAt: String? = null
)

data class LeadDto(
    val _id: String,
    val name: String,
    val email: String? = null,
    val phone: String? = null,
    val phoneNumber: String? = null,
    val company: String? = null,
    val source: String = "manual",
    val status: String = "New",
    val dealValue: Double? = null,
    val currency: String? = "INR",
    val score: Int? = null,
    val notes: List<LeadNoteDto>? = null,
    val tags: List<String>? = null,
    val assignedAgentId: Any? = null,
    val assignedAgentName: String? = null,
    val conversationId: String? = null,
    val visitorId: String? = null,
    val createdAt: String? = null,
    val updatedAt: String? = null
) {
    val displayPhone: String?
        get() = phone ?: phoneNumber
}

data class LeadsResponse(
    val leads: List<LeadDto>,
    val total: Int? = null,
    val page: Int? = null,
    val pages: Int? = null
)

data class CreateLeadRequest(
    val name: String,
    val email: String? = null,
    val phone: String? = null,
    val company: String? = null,
    val source: String? = "manual",
    val status: String? = "New",
    val dealValue: Double? = null,
    val currency: String? = "USD",
    val score: Int? = 50,
    val notes: List<String>? = null,
    val tags: List<String>? = null,
    val assignedAgentId: String? = null,
    val conversationId: String? = null,
    val visitorId: String? = null
)

data class LeadStatsDto(
    val totalLeads: Int = 0,
    val newLeads: Int = 0,
    val wonLeads: Int = 0,
    val lostLeads: Int = 0,
    val totalPipelineValue: Double = 0.0,
    val wonValue: Double = 0.0,
    val conversionRate: Double = 0.0
)

// ============================================
// RETROFIT API INTERFACES
// ============================================
interface LetsTrackApi {
    @POST("/api/auth/login")
    suspend fun login(@Body request: LoginRequest): LoginResponse

    @POST("/api/auth/google-login")
    suspend fun googleLogin(@Body request: GoogleLoginRequest): LoginResponse

    @POST("/api/auth/reset-password")
    suspend fun resetPassword(@Body request: ResetPasswordRequest): ResetPasswordResponse

    @GET("/api/analytics/summary")
    suspend fun getAnalytics(@Header("Authorization") token: String): AnalyticsResponse

    @GET("/api/conversations/{id}/messages")
    suspend fun getMessages(
        @Header("Authorization") token: String,
        @Path("id") conversationId: String
    ): List<MessageDto>

    @POST("/api/auth/fcm-token")
    suspend fun registerFcmToken(
        @Header("Authorization") token: String,
        @Body request: FcmTokenRequest
    ): retrofit2.Response<Unit>

    @GET("/api/quick-replies")
    suspend fun getQuickReplies(@Header("Authorization") token: String): List<QuickReplyDto>

    @PUT("/api/auth/profile")
    suspend fun updateProfile(
        @Header("Authorization") token: String,
        @Body request: UpdateProfileRequest
    ): UserProfile

    @POST("/api/auth/register-agent")
    suspend fun registerAgent(
        @Header("Authorization") token: String,
        @Body request: RegisterAgentRequest
    ): RegisterAgentResponse

    @PUT("/api/visitors/{id}")
    suspend fun updateVisitor(
        @Header("Authorization") token: String,
        @Path("id") visitorId: String,
        @Body body: Map<String, @JvmSuppressWildcards Any>
    ): VisitorDto

    @GET("/api/visitors/{id}")
    suspend fun getVisitor(
        @Header("Authorization") token: String,
        @Path("id") visitorId: String
    ): VisitorDto

    // Lead Management APIs
    @GET("/api/leads")
    suspend fun getLeads(
        @Header("Authorization") token: String,
        @Query("search") search: String? = null,
        @Query("status") status: String? = null,
        @Query("source") source: String? = null
    ): LeadsResponse

    @POST("/api/leads")
    suspend fun createLead(
        @Header("Authorization") token: String,
        @Body request: CreateLeadRequest
    ): LeadDto

    @PUT("/api/leads/{id}")
    suspend fun updateLead(
        @Header("Authorization") token: String,
        @Path("id") leadId: String,
        @Body body: Map<String, @JvmSuppressWildcards Any>
    ): LeadDto

    @GET("/api/leads/stats")
    suspend fun getLeadStats(
        @Header("Authorization") token: String
    ): LeadStatsDto

    @POST("/api/leads/{id}/notes")
    suspend fun addLeadNote(
        @Header("Authorization") token: String,
        @Path("id") leadId: String,
        @Body body: Map<String, String>
    ): LeadDto
}

// ============================================
// UNIFIED NETWORK CLIENT MANAGER
// ============================================
object NetworkClient {
    private const val BASE_URL = "https://letstrack.manacity.in"

    private var authToken: String? = null
    var currentUser: UserProfile? = null
    var currentTenant: TenantDetails? = null
    var cachedAgents: List<UserProfile> = emptyList()

    val gson = Gson()
    
    // Retrofit service
    val api: LetsTrackApi by lazy {
        val logging = HttpLoggingInterceptor().apply {
            level = HttpLoggingInterceptor.Level.BODY
        }
        val client = OkHttpClient.Builder()
            .addInterceptor(logging)
            .build()

        Retrofit.Builder()
            .baseUrl(BASE_URL)
            .client(client)
            .addConverterFactory(GsonConverterFactory.create())
            .build()
            .create(LetsTrackApi::class.java)
    }

    // Socket.io instance
    private var socket: Socket? = null

    fun setAuth(token: String, user: UserProfile, tenant: TenantDetails) {
        authToken = "Bearer $token"
        currentUser = user
        currentTenant = tenant
    }

    fun getAuthHeader(): String = authToken ?: ""

    @Synchronized
    fun getSocketInstance(): Socket {
        if (socket == null) {
            try {
                val opts = IO.Options().apply {
                    forceNew = true
                    reconnection = true
                }
                // Connect to `/dashboard` namespace
                socket = IO.socket("$BASE_URL/dashboard", opts)
            } catch (e: URISyntaxException) {
                e.printStackTrace()
            }
        }
        return socket!!
    }

    fun connectSocket() {
        val s = getSocketInstance()
        if (!s.connected()) {
            s.connect()
            
            // On connect, authenticate
            s.on(Socket.EVENT_CONNECT) {
                val initData = JSONObject().apply {
                    put("tenantId", currentTenant?.id)
                    put("agentId", currentUser?.id)
                }
                s.emit("agent-init", initData)
            }
        } else {
            // Already connected, immediately emit agent-init to ensure lists sync!
            val initData = JSONObject().apply {
                put("tenantId", currentTenant?.id)
                put("agentId", currentUser?.id)
            }
            s.emit("agent-init", initData)
        }
    }

    fun disconnectSocket() {
        socket?.disconnect()
        socket = null
    }
}

File: ios/LetsTrack/NetworkClient.swift
import Foundation
import Combine

final class NetworkClient: ObservableObject {
    static let shared = NetworkClient()
    
    private let baseURL = "https://letstrack.manacity.in"
    
    @Published var currentUser: UserProfile?
    @Published var currentTenant: TenantDetails?
    @Published var authToken: String?
    @Published var cachedAgents: [UserProfile] = []
    
    private let userDefaults = UserDefaults.standard
    private let decoder = JSONDecoder()
    private let encoder = JSONEncoder()
    
    private init() {
        // Load saved session on initialization
        if let token = userDefaults.string(forKey: "auth_token"),
           let userJson = userDefaults.data(forKey: "user_profile"),
           let tenantJson = userDefaults.data(forKey: "tenant_details") {
            do {
                let user = try decoder.decode(UserProfile.self, from: userJson)
                let tenant = try decoder.decode(TenantDetails.self, from: tenantJson)
                self.authToken = token
                self.currentUser = user
                self.currentTenant = tenant
                
                // Re-register FCM token on startup if we have it saved
                if let fcmToken = userDefaults.string(forKey: "fcm_token") {
                    Task {
                        do {
                            try await self.registerFcmToken(fcmToken: fcmToken)
                            print("[Push Notification Debug] FCM Token re-registered successfully on launch.")
                        } catch {
                            print("[Push Notification Debug] Failed to re-register FCM Token on launch: \(error)")
                        }
                    }
                }
            } catch {
                print("[Session Decode Debug] Failed to decode saved profile sessions with error: \(error)")
                clearAuth()
            }
        }
    }
    
    var isAuthenticated: Bool {
        authToken != nil && currentUser != nil && currentTenant != nil
    }
    
    func setAuth(token: String, user: UserProfile, tenant: TenantDetails) {
        self.authToken = token
        self.currentUser = user
        self.currentTenant = tenant
        
        userDefaults.set(token, forKey: "auth_token")
        if let userJson = try? encoder.encode(user) {
            userDefaults.set(userJson, forKey: "user_profile")
        }
        if let tenantJson = try? encoder.encode(tenant) {
            userDefaults.set(tenantJson, forKey: "tenant_details")
        }
        
        // Register saved FCM token if available
        if let fcmToken = userDefaults.string(forKey: "fcm_token") {
            Task {
                do {
                    try await self.registerFcmToken(fcmToken: fcmToken)
                    print("[Push Notification Debug] FCM Token registered successfully with backend on login.")
                } catch {
                    print("[Push Notification Debug] Failed to register FCM Token with backend on login: \(error)")
                }
            }
        }
    }
    
    func clearAuth() {
        self.authToken = nil
        self.currentUser = nil
        self.currentTenant = nil
        
        userDefaults.removeObject(forKey: "auth_token")
        userDefaults.removeObject(forKey: "user_profile")
        userDefaults.removeObject(forKey: "tenant_details")
        
        // Disconnect sockets
        SocketManager.shared.disconnectSocket()
    }
    
    func getAuthHeader() -> String {
        guard let token = authToken else { return "" }
        return "Bearer \(token)"
    }
    
    // MARK: - API Calls
    
    func login(request: LoginRequest) async throws -> LoginResponse {
        let url = URL(string: "\(baseURL)/api/auth/login")!
        var urlRequest = URLRequest(url: url)
        urlRequest.httpMethod = "POST"
        urlRequest.setValue("application/json", forHTTPHeaderField: "Content-Type")
        urlRequest.httpBody = try encoder.encode(request)
        
        let (data, response) = try await URLSession.shared.data(for: urlRequest)
        guard let httpResponse = response as? HTTPURLResponse, (200...299).contains(httpResponse.statusCode) else {
            throw NSError(domain: "NetworkClient", code: 401, userInfo: [NSLocalizedDescriptionKey: "Invalid credentials or network failure."])
        }
        
        let loginResponse = try decoder.decode(LoginResponse.self, from: data)
        
        // Update state on Main thread
        await MainActor.run {
            self.setAuth(token: loginResponse.token, user: loginResponse.user, tenant: loginResponse.tenant)
        }
        
        return loginResponse
    }
    
    func googleLogin(idToken: String) async throws -> LoginResponse {
        let url = URL(string: "\(baseURL)/api/auth/google-login")!
        var urlRequest = URLRequest(url: url)
        urlRequest.httpMethod = "POST"
        urlRequest.setValue("application/json", forHTTPHeaderField: "Content-Type")
        let requestPayload = GoogleLoginRequest(credential: idToken)
        urlRequest.httpBody = try encoder.encode(requestPayload)
        
        let (data, response) = try await URLSession.shared.data(for: urlRequest)
        guard let httpResponse = response as? HTTPURLResponse else {
            throw NSError(domain: "NetworkClient", code: 500, userInfo: [NSLocalizedDescriptionKey: "Server response error."])
        }
        
        if !(200...299).contains(httpResponse.statusCode) {
            if let errorJson = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
               let message = errorJson["error"] as? String {
                throw NSError(domain: "NetworkClient", code: httpResponse.statusCode, userInfo: [NSLocalizedDescriptionKey: message])
            }
            throw NSError(domain: "NetworkClient", code: httpResponse.statusCode, userInfo: [NSLocalizedDescriptionKey: "Google Sign-In failed on server."])
        }
        
        let loginResponse = try decoder.decode(LoginResponse.self, from: data)
        
        await MainActor.run {
            self.setAuth(token: loginResponse.token, user: loginResponse.user, tenant: loginResponse.tenant)
        }
        
        return loginResponse
    }
    
    func resetPassword(request: ResetPasswordRequest) async throws -> ResetPasswordResponse {
        let url = URL(string: "\(baseURL)/api/auth/reset-password")!
        var urlRequest = URLRequest(url: url)
        urlRequest.httpMethod = "POST"
        urlRequest.setValue("application/json", forHTTPHeaderField: "Content-Type")
        urlRequest.httpBody = try encoder.encode(request)
        
        let (data, response) = try await URLSession.shared.data(for: urlRequest)
        guard let httpResponse = response as? HTTPURLResponse, (200...299).contains(httpResponse.statusCode) else {
            throw NSError(domain: "NetworkClient", code: 400, userInfo: [NSLocalizedDescriptionKey: "Password reset failed."])
        }
        
        return try decoder.decode(ResetPasswordResponse.self, from: data)
    }
    
    func getAnalytics() async throws -> AnalyticsResponse {
        let url = URL(string: "\(baseURL)/api/analytics/summary")!
        var urlRequest = URLRequest(url: url)
        urlRequest.httpMethod = "GET"
        urlRequest.setValue(getAuthHeader(), forHTTPHeaderField: "Authorization")
        
        let (data, _) = try await URLSession.shared.data(for: urlRequest)
        return try decoder.decode(AnalyticsResponse.self, from: data)
    }
    
    func getMessages(conversationId: String) async throws -> [MessageDto] {
        let url = URL(string: "\(baseURL)/api/conversations/\(conversationId)/messages")!
        var urlRequest = URLRequest(url: url)
        urlRequest.httpMethod = "GET"
        urlRequest.setValue(getAuthHeader(), forHTTPHeaderField: "Authorization")
        
        let (data, _) = try await URLSession.shared.data(for: urlRequest)
        return try decoder.decode([MessageDto].self, from: data)
    }
    
    func registerFcmToken(fcmToken: String) async throws {
        let url = URL(string: "\(baseURL)/api/auth/fcm-token")!
        var urlRequest = URLRequest(url: url)
        urlRequest.httpMethod = "POST"
        urlRequest.setValue("application/json", forHTTPHeaderField: "Content-Type")
        urlRequest.setValue(getAuthHeader(), forHTTPHeaderField: "Authorization")
        urlRequest.httpBody = try encoder.encode(FcmTokenRequest(fcmToken: fcmToken))
        
        _ = try await URLSession.shared.data(for: urlRequest)
    }
    
    func getQuickReplies() async throws -> [QuickReplyDto] {
        let url = URL(string: "\(baseURL)/api/quick-replies")!
        var urlRequest = URLRequest(url: url)
        urlRequest.httpMethod = "GET"
        urlRequest.setValue(getAuthHeader(), forHTTPHeaderField: "Authorization")
        
        let (data, _) = try await URLSession.shared.data(for: urlRequest)
        return try decoder.decode([QuickReplyDto].self, from: data)
    }
    
    func updateProfile(request: UpdateProfileRequest) async throws -> UserProfile {
        let url = URL(string: "\(baseURL)/api/auth/profile")!
        var urlRequest = URLRequest(url: url)
        urlRequest.httpMethod = "PUT"
        urlRequest.setValue("application/json", forHTTPHeaderField: "Content-Type")
        urlRequest.setValue(getAuthHeader(), forHTTPHeaderField: "Authorization")
        urlRequest.httpBody = try encoder.encode(request)
        
        let (data, _) = try await URLSession.shared.data(for: urlRequest)
        let updatedUser = try decoder.decode(UserProfile.self, from: data)
        
        await MainActor.run {
            self.currentUser = updatedUser
            if let userJson = try? encoder.encode(updatedUser) {
                userDefaults.set(userJson, forKey: "user_profile")
            }
        }
        
        return updatedUser
    }
    
    func registerAgent(request: RegisterAgentRequest) async throws -> RegisterAgentResponse {
        let url = URL(string: "\(baseURL)/api/auth/register-agent")!
        var urlRequest = URLRequest(url: url)
        urlRequest.httpMethod = "POST"
        urlRequest.setValue("application/json", forHTTPHeaderField: "Content-Type")
        urlRequest.setValue(getAuthHeader(), forHTTPHeaderField: "Authorization")
        urlRequest.httpBody = try encoder.encode(request)
        
        let (data, response) = try await URLSession.shared.data(for: urlRequest)
        guard let httpResponse = response as? HTTPURLResponse, (200...299).contains(httpResponse.statusCode) else {
            throw NSError(domain: "NetworkClient", code: 400, userInfo: [NSLocalizedDescriptionKey: "Registration failed. Email might already exist."])
        }
        
        return try decoder.decode(RegisterAgentResponse.self, from: data)
    }
    
    func getVisitor(visitorId: String) async throws -> VisitorDto {
        let url = URL(string: "\(baseURL)/api/visitors/\(visitorId)")!
        var urlRequest = URLRequest(url: url)
        urlRequest.httpMethod = "GET"
        urlRequest.setValue(getAuthHeader(), forHTTPHeaderField: "Authorization")
        
        let (data, _) = try await URLSession.shared.data(for: urlRequest)
        return try decoder.decode(VisitorDto.self, from: data)
    }
    
    func updateVisitor(visitorId: String, fields: [String: Any]) async throws -> VisitorDto {
        let url = URL(string: "\(baseURL)/api/visitors/\(visitorId)")!
        var urlRequest = URLRequest(url: url)
        urlRequest.httpMethod = "PUT"
        urlRequest.setValue("application/json", forHTTPHeaderField: "Content-Type")
        urlRequest.setValue(getAuthHeader(), forHTTPHeaderField: "Authorization")
        
        let jsonSerialization = try JSONSerialization.data(withJSONObject: fields)
        urlRequest.httpBody = jsonSerialization
        
        let (data, _) = try await URLSession.shared.data(for: urlRequest)
        return try decoder.decode(VisitorDto.self, from: data)
    }
    
    // MARK: - Lead Management APIs
    
    func getLeads(search: String? = nil, status: String? = nil, source: String? = nil) async throws -> [LeadDto] {
        var urlComponents = URLComponents(string: "\(baseURL)/api/leads")!
        var queryItems: [URLQueryItem] = []
        if let search = search, !search.isEmpty {
            queryItems.append(URLQueryItem(name: "search", value: search))
        }
        if let status = status, status != "All" && !status.isEmpty {
            queryItems.append(URLQueryItem(name: "status", value: status))
        }
        if let source = source, source != "All" && !source.isEmpty {
            queryItems.append(URLQueryItem(name: "source", value: source))
        }
        if !queryItems.isEmpty {
            urlComponents.queryItems = queryItems
        }
        
        var urlRequest = URLRequest(url: urlComponents.url!)
        urlRequest.httpMethod = "GET"
        urlRequest.setValue(getAuthHeader(), forHTTPHeaderField: "Authorization")
        
        let (data, response) = try await URLSession.shared.data(for: urlRequest)
        guard let httpResponse = response as? HTTPURLResponse, (200...299).contains(httpResponse.statusCode) else {
            throw NSError(domain: "NetworkClient", code: 400, userInfo: [NSLocalizedDescriptionKey: "Failed to fetch leads."])
        }
        
        // Backend returns { leads: [...], total, page, pages }
        struct LeadsResponse: Codable {
            let leads: [LeadDto]
        }
        
        if let resp = try? decoder.decode(LeadsResponse.self, from: data) {
            return resp.leads
        }
        return try decoder.decode([LeadDto].self, from: data)
    }
    
    func createLead(request: CreateLeadRequest) async throws -> LeadDto {
        let url = URL(string: "\(baseURL)/api/leads")!
        var urlRequest = URLRequest(url: url)
        urlRequest.httpMethod = "POST"
        urlRequest.setValue("application/json", forHTTPHeaderField: "Content-Type")
        urlRequest.setValue(getAuthHeader(), forHTTPHeaderField: "Authorization")
        urlRequest.httpBody = try encoder.encode(request)
        
        let (data, response) = try await URLSession.shared.data(for: urlRequest)
        guard let httpResponse = response as? HTTPURLResponse, (200...299).contains(httpResponse.statusCode) else {
            throw NSError(domain: "NetworkClient", code: 400, userInfo: [NSLocalizedDescriptionKey: "Failed to create lead."])
        }
        
        return try decoder.decode(LeadDto.self, from: data)
    }
    
    func updateLead(leadId: String, fields: [String: Any]) async throws -> LeadDto {
        let url = URL(string: "\(baseURL)/api/leads/\(leadId)")!
        var urlRequest = URLRequest(url: url)
        urlRequest.httpMethod = "PUT"
        urlRequest.setValue("application/json", forHTTPHeaderField: "Content-Type")
        urlRequest.setValue(getAuthHeader(), forHTTPHeaderField: "Authorization")
        urlRequest.httpBody = try JSONSerialization.data(withJSONObject: fields)
        
        let (data, response) = try await URLSession.shared.data(for: urlRequest)
        guard let httpResponse = response as? HTTPURLResponse, (200...299).contains(httpResponse.statusCode) else {
            throw NSError(domain: "NetworkClient", code: 400, userInfo: [NSLocalizedDescriptionKey: "Failed to update lead."])
        }
        
        return try decoder.decode(LeadDto.self, from: data)
    }
    
    func getLeadStats() async throws -> LeadStatsDto {
        let url = URL(string: "\(baseURL)/api/leads/stats")!
        var urlRequest = URLRequest(url: url)
        urlRequest.httpMethod = "GET"
        urlRequest.setValue(getAuthHeader(), forHTTPHeaderField: "Authorization")
        
        let (data, response) = try await URLSession.shared.data(for: urlRequest)
        guard let httpResponse = response as? HTTPURLResponse, (200...299).contains(httpResponse.statusCode) else {
            throw NSError(domain: "NetworkClient", code: 400, userInfo: [NSLocalizedDescriptionKey: "Failed to fetch lead stats."])
        }
        
        return try decoder.decode(LeadStatsDto.self, from: data)
    }
    
    func addLeadNote(leadId: String, text: String) async throws -> LeadDto {
        let url = URL(string: "\(baseURL)/api/leads/\(leadId)/notes")!
        var urlRequest = URLRequest(url: url)
        urlRequest.httpMethod = "POST"
        urlRequest.setValue("application/json", forHTTPHeaderField: "Content-Type")
        urlRequest.setValue(getAuthHeader(), forHTTPHeaderField: "Authorization")
        urlRequest.httpBody = try JSONSerialization.data(withJSONObject: ["text": text])
        
        let (data, response) = try await URLSession.shared.data(for: urlRequest)
        guard let httpResponse = response as? HTTPURLResponse, (200...299).contains(httpResponse.statusCode) else {
            throw NSError(domain: "NetworkClient", code: 400, userInfo: [NSLocalizedDescriptionKey: "Failed to add note to lead."])
        }
        
        return try decoder.decode(LeadDto.self, from: data)
    }
}
